### PR TITLE
Adding checksum to all write characteristics

### PR DESCRIPTION
**Pull request by Alka1703**
Visit the exception throwing of CRC mismatch during the VC transfer before merging

### CODE REVIEW

File: android/src/main/java/io/mosip/tuvali/ble/peripheral/Peripheral.kt [hunk 1]
```diff
@@ -36,7 +38,13 @@ class Peripheral(context: Context, peripheralListener: IPeripheralListener) {
   }
 
   private fun notifyDisconnect(serviceUUID: UUID) {
-    sendData(serviceUUID, GattService.DISCONNECT_CHAR_UUID, byteArrayOf(Verifier.DISCONNECT_STATUS.toByte()))
+    val data = byteArrayOf(Verifier.DISCONNECT_STATUS.toByte())
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/ble/peripheral/Peripheral.kt:41]: I think we should refactor on develop. Peripheral layer being the lower layer should avoid having business logic. It should just send data to whatever characteristic given by higher layers.

At the same time, adding checksum data can be encapsulated into DisconnectCharData class.

File: android/src/main/java/io/mosip/tuvali/transfer/Assembler.kt [hunk 2]
```diff
@@ -64,7 +55,7 @@ class Assembler(private val totalSize: Int, private val maxDataBytes: Int ): Chu
     var missedSeqNumbers = intArrayOf()
     chunkReceivedMarker.forEachIndexed() { i, elem ->
       if (elem != chunkReceivedMarkerByte) {
-        //Log.d(logTag, "getMissedSequenceNumbers: adding missed sequence number $i")
+        Log.d(logTag, "getMissedSequenceNumbers: adding missed sequence number $i")
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/transfer/Assembler.kt]: comment out this

File: android/src/main/java/io/mosip/tuvali/transfer/CRCValidator.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    val chunkData = data.dropLast(2).toByteArray()
+    val calculatedCRC = calculate(chunkData)
+    if(calculatedCRC != receivedCRC) {
+      Log.e(logTag, "CRC check failed for $characteristic. Received CRC: $receivedCRC, Calculated CRC: $calculatedCRC")
+      return false
+    }
+    return true
+  }
+
+  fun validateCrcSentByWallet(data: ByteArray, characteristic: String): Boolean{
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/transfer/CRCValidator.kt]: validateDataFromWallet

File: android/src/main/java/io/mosip/tuvali/transfer/CRCValidator.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    if(!verify(data, characteristic))
+      throw io.mosip.tuvali.verifier.exception.CrcValidationFailedException(characteristic)
+    return true
+  }
+
+  fun validateCrcSentByVerifier(data: ByteArray, characteristic: String): Boolean{
+    if(!verify(data, characteristic))
+      throw io.mosip.tuvali.wallet.exception.CrcValidationFailedException(characteristic)
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/transfer/CRCValidator.kt]: Exception can have all the details like receivedCrc, length of chunk data, calculatedCrc etc

File: android/src/main/java/io/mosip/tuvali/transfer/Chunker.kt [hunk 5]
```diff
@@ -3,6 +3,8 @@ package io.mosip.tuvali.transfer
 import android.util.Log
 import io.mosip.tuvali.transfer.Util.Companion.intToTwoBytesBigEndian
 import io.mosip.tuvali.transfer.Util.Companion.getLogTag
+import org.bouncycastle.util.encoders.Hex
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/transfer/Chunker.kt]: optimise imports

File: android/src/main/java/io/mosip/tuvali/transfer/Chunker.kt [hunk 6]
```diff
@@ -44,24 +45,24 @@ class Chunker(private val data: ByteArray, private val maxDataBytes: Int) :
...
 
   /*
-  <--------------------------------------------------Max Data Bytes -------------------------------------------------------------->
-  +-----------------------+-----------------------------+-------------------------------------------------------------------------+
-  |                       |                             |                                                                         |
-  |  chunk sequence no    |   checksum value of data    |         chunk payload                                                   |
-  |      (2 bytes)        |         (2 bytes)           |       (upto MaxDataBytes -4 bytes)                                      |
-  |                       |                             |                                                                         |
-  +-----------------------+-----------------------------+-------------------------------------------------------------------------+
+      <------------------------------- MTU ----------------------------------------------->
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/transfer/Chunker.kt]: MTU can be changed to Max Chunk Data Bytes

File: android/src/main/java/io/mosip/tuvali/transfer/Util.kt [hunk 7]
```diff
@@ -9,6 +9,7 @@ import java.security.MessageDigest
 import java.util.zip.GZIPInputStream
 import java.util.zip.GZIPOutputStream
 import io.mosip.tuvali.openid4vpble.Openid4vpBleModule
+import java.nio.ByteBuffer
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/transfer/Util.kt:12]: optimise imports

File: android/src/main/java/io/mosip/tuvali/verifier/GattService.kt [hunk 8]
```diff
@@ -7,13 +7,61 @@ import java.util.*
...
+    /*
+       +---------------------------------- + ----------------------------------- +
+       |                                   |                                     |
+       |           Verification            |   Checksum value of Response size   |
+       |    Accepted/Rejected(1 byte)      |        (2 bytes)                    |
+       |                                   |                                     |
+       +---------------------------------- + ----------------------------------- +
+    */
     val VERIFICATION_STATUS_CHAR_UUID: UUID = UUID.fromString("00002037-0000-1000-8000-00805f9b34fb")
     val DISCONNECT_CHAR_UUID: UUID = UUID.fromString("00002038-0000-1000-8000-00805f9b34fb")
```
Comment (krishnakumar4a4) [inline android/src/main/java/io/mosip/tuvali/verifier/GattService.kt:74]: add the ascii art

File: ios/crc/CRC16.swift [hunk 9]
```diff
@@ -18,12 +18,15 @@ class CRC {
...
-        let got = evaluate(d: d)
-        if got == expected {
-            return true
+    static func verify(data: Data, characteristic: String) -> Bool {
+        let crcValueReceived = Utils.twoBytesToIntBigEndian(num: Data(data.suffix(2)))
+        let crcValueCalculated = calculate(d: data.dropLast(2))
+
+        if crcValueCalculated != crcValueReceived {
+            os_log(.error, "CRC check failed. Received CRC: %{public}d, Calculated CRC: %{public}d", crcValueReceived, crcValueCalculated)
+            ErrorHandler.sharedInstance.handle(error: OpenId4vpError.crcCheckFailedError(characteristic: characteristic))
```
Comment (vharsh) [inline ios/crc/CRC16.swift:27]:
```suggestion
```

IMO, this would throw an error and halt the share, I believe this shouldn't be an expected behaviour from a Utils class, i.e. this line must be omitted.

File: ios/crc/CRC16.swift [hunk 10]
```diff
@@ -18,12 +18,15 @@ class CRC {
         return crc
     }
 
-    static func verify(d: Data, expected: UInt16) -> Bool {
-        let got = evaluate(d: d)
-        if got == expected {
-            return true
+    static func verify(data: Data, characteristic: String) -> Bool {
```
Comment (vharsh) [inline ios/crc/CRC16.swift:21]: (nitpick) IMO CRC is a concept and BLE transfer is a concept, we might not want to mix them together in the lower level classes and leave these for the business logic layer to figure out based on this function's boolean return value. 


Tl; dr: This class deals with CRC, it shouldn't ideally deal with the logic of what is being compared.

File: ios/ble/Utility/TransferHandler.swift [hunk 11]
```diff
@@ -77,32 +77,35 @@ class TransferHandler {
 
     private func requestTransmissionReport() {
         var notifyObj: Data
-        delegate?.write(serviceUuid: BLEConstants.SERVICE_UUID, charUUID: NetworkCharNums.TRANSFER_REPORT_REQUEST_CHAR_UUID, data: withUnsafeBytes(of: 1.littleEndian) { Data($0) }, withResponse: true)
+        let data  = withUnsafeBytes(of: 1.littleEndian) { Data($0) }
```
Comment (vharsh) [inline ios/ble/Utility/TransferHandler.swift:80]: Was this fixed earlier using some other standard library call? I'll try to find the replacement code snippet for this logic

File: ios/Wallet/characteristics/VerificationStatusCharacteristic.swift [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+class VerificationStatusCharacteristic {
+  var data: Data
+  static let uuid = CharacteristicIds.VERIFICATION_STATUS_CHAR_UUID
+  var status : Int
+
+  init(data: Data){
+      self.data =  data
+      status = Int(data[0])
+       CRCValidator.verify(data: data, characteristic: VerificationStatusCharacteristic.uuid.rawValue)
```
Comment (vharsh) [inline ios/Wallet/characteristics/VerificationStatusCharacteristic.swift:14]: (nit) fix code formatting

File: ios/Wallet/characteristics/DisconnectCharacteristic.swift [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+class DisconnectCharacteristic {
+  var data: Data
+  static let uuid = CharacteristicIds.DISCONNECT_CHAR_UUID
+  var status:  Int
+
+    init(data: Data){
+        self.data =  data
+        status = Int(data[0])
+         CRCValidator.verify(data: data, characteristic: DisconnectCharacteristic.uuid.rawValue)
```
Comment (vharsh) [inline ios/Wallet/characteristics/DisconnectCharacteristic.swift:14]: I now sort of understand why an error is thrown from CRCValidator, but have you tried to see if an error doesn't halt the transfer?

**Review by vharsh — COMMENTED** [review]
did a partial review from below file changes to explicitly look into and understand changes done on Swiftlang side for iOS; got some questions in CRCValidation & it's coupling with ErrorHandler, please see if this is okay & **does not hamper user flow in due course of a couple of bit-flips happening over the network**
